File: greed/LC6932.cpp
#include <bits/stdc++.h>
// https://leetcode.cn/problems/maximum-elegance-of-a-k-length-subsequence/

using namespace std;
#define i64 long long
class Solution {
public:
    long long findMaximumElegance(vector<vector<int>>& items, int k) {
        sort(items.begin(), items.end(), [&](const auto &item1, const auto &item2) {
            return item1[0] > item2[0];
        });
        i64 total_profit = 0;
        i64 ans = 0;
        int n = items.size();
        unordered_set<int> vis;
        stack<int> duplicates;
        for (int i=0;i<n;++i){
            int profit = items[i][0];
            int category = items[i][1];
            if (i+1 <= k) {
                total_profit += profit;
                if (!vis.insert(category).second) {
                    duplicates.push(profit);
                }
            } else if(!duplicates.empty() && vis.insert(category).second) {
                total_profit -= duplicates.top();
                total_profit += profit;
                duplicates.pop();
            }
            i64 cnt = vis.size();
            ans = max(ans, cnt*cnt + total_profit);
        }
        return ans;
    }
};

File: greed/LC6932.py
# https://leetcode.cn/problems/maximum-elegance-of-a-k-length-subsequence/
from typing import *


class Solution:
    def findMaximumElegance(self, items: List[List[int]], k: int) -> int:
        items.sort(key=lambda x: -x[0])  # profit max~min
        ans = total_profit = 0
        vis = set()
        duplicates = []
        for i, (profit, category) in enumerate(items):
            if i+1 <= k:
                total_profit += profit
                if category in vis:
                    duplicates.append(profit)
                else:
                    vis.add(category)
            elif category not in vis and duplicates:
                # profit already sorted so only consider new item if it's a new category AND there're duplicates in any existing category
                vis.add(category)
                total_profit -= duplicates.pop()
                total_profit += profit
            ans = max(ans, total_profit + len(vis)**2)
        return ans
